packages/esm-patient-orders-app/src/lab-results/lab-results.resource.ts

```diff
@@ -157,3 +158,69 @@ export async function updateOrderResult(
   }
   throw new Error('Failed to update order');
 }
+
+export function createObservationPayload(concept: LabOrderConcept, order: Order, values: any) {
+  if (concept.set && concept.setMembers.length > 0) {
+    const groupMembers = concept.setMembers
+      .map((member) => createGroupMember(member, order, values))
+      .filter((member) => member?.value !== null && member?.value !== undefined);
+
+    if (groupMembers.length === 0) {
+      return { obs: [] };
+    }
+
+    return { obs: [createObservation(order, groupMembers)] };
+  }
+
+  if (!concept.set && concept.setMembers.length === 0) {
+    const value = getValue(concept, values);
+    if (value === null || value === undefined) {
+      return { obs: [] };
+    }
+    return { obs: [createObservation(order, null, value)] };
+  }
+
+  return { obs: [] };
+}
+
+function createGroupMember(member, order, values) {
+  const value = getValue(member, values);
+  if (value === null || value === undefined) {
+    return null;
+  }
+  return {
+    concept: { uuid: member.uuid },
+    value: value,
+    status: 'FINAL',
+    order: { uuid: order.uuid },
+  };
+}
+
+function createObservation(order, groupMembers = null, value = null) {
+  return {
+    concept: { uuid: order.concept.uuid },
+    status: 'FINAL',
+    order: { uuid: order.uuid },
+    ...(groupMembers && groupMembers.length > 0 && { groupMembers }),
+    ...(value !== null && value !== undefined && { value }),
+  };
+}
+
+function getValue(concept, values) {
+  const { datatype, uuid } = concept;
+  const value = values[uuid];
+
+  if (value === null || value === undefined) {
+    return null;
+  }
+
+  if (['Numeric', 'Text'].includes(datatype.display)) {
```

Comment:
This is actually a case where using `display` is incorrect. `display` is a user-facing property because its localizable, but because its localizable, we cannot rely on it's value being English. In this case, the "correct" value is (for historical reasons) `hl7Abbreviation`, but `name` is less likely to be wrong.

Reply:
`hl7Abbreviation` is something like:

|Datatype|Code|
|---|---|
|Coded|CWE|
|Text|ST|
|Numeric|NM|